### PR TITLE
Added report template generation

### PR DESCRIPTION
**Pull request by furkacz**

### CODE REVIEW

**Review by dlydka — DISMISSED** [review]

File: src/report/report_generate.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+"""
+Script generating monthly report
+"""
+
+import report_utils as utils
+import sys
+from report_data import get_report_data
+from docx import Document
+from docx.enum import section, text
+from os import remove
```
Comment (BognaLew) [inline src/report/report_generate.py]: Imports should be grouped in the following order:
1. Standard library imports.
2. Related third party imports.
3. Local application/library specific imports.
You should put a blank line between each group of imports.

Comment (BognaLew) [reply src/report/report_generate.py]: https://peps.python.org/pep-0008/#imports

File: src/report/report_generate.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+
```
Comment (BognaLew) [inline src/report/report_generate.py]: why the code isn't in a function?

File: src/report/report_generate.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        summary_table.cell(1, index + 1).paragraphs[0].text = f'{data["recent"][key]["subscribed"]}'
+        summary_table.cell(2, index + 1).paragraphs[0].text = f'{data["recent"][key]["cancelled"]}'
+
+    document.save(f'{directory}report-{data["date"]}.docx')
+
+finally:
+    # image cleanup
+    remove(users_chart)
+    remove(users_age_chart)
+    remove(users_gender_chart)
```
Comment (BognaLew) [inline src/report/report_generate.py]: you can `if __name__ == "__main__":` at the end of a file and call a function with this code

File: src/report/report_generate.py (hunk 2 above)
Comment (furkacz) [reply src/report/report_generate.py]: I'm going to change it to be a function in this sprint

**Review by BognaLew — CHANGES_REQUESTED** [review]
add unit tests

File: src/report/report_generate.py (hunk 1 above)
Comment (furkacz) [reply src/report/report_generate.py]: I'll fix it in the current sprint

File: src/report/report_generate.py (hunk 3 above)
Comment (furkacz) [reply src/report/report_generate.py]: I've already did this in the current sprint

File: src/report/report_generate.py (hunk 2 above)
Comment (BognaLew) [reply src/report/report_generate.py]: This should have been done in a previous sprint. Update changes to this pull request.

File: src/report/report_data.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            'silver': {
+                'subscribed': 500,
+                'cancelled': 150,
+            },
+            'gold': {
+                'subscribed': 200,
+                'cancelled': 100,
+            }
+        }
+    }
```
Comment (BognaLew) [inline src/report/report_data.py:58]: This shouldn't be in src package.

File: requirements.txt [hunk 5]
```diff
@@ -13,4 +13,8 @@ pytest==7.4.2
 Werkzeug==3.0.0
 Flask-WTF==1.2.1
 Bootstrap-Flask==2.3.2
-email_validator==2.0.0
\ No newline at end of file
+email_validator==2.0.0
+python-docx==1.0.1
+typing-extensions==4.8.0
+lxml==4.9.3
+matplotlib==3.5.3
```
Comment (BognaLew) [inline requirements.txt]: Why are you using `matplotlib==3.5.3` when there are newer versions?

File: src/report/report.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+from docx.enum import section, text
+from docx.shared import RGBColor
+
+import report_utils as utils
+import report_charts as charts
+from report_data import get_report_data
+
+
+def generate(directory: str = '') -> str:
```
Comment (BognaLew) [inline src/report/report.py:17]: Shouldn't it be splitted into smaller functions?

File: src/report/report_utils.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+from docx import Document
+from docx.text.paragraph import Paragraph
+from docx.table import Table, _Cell
+from docx.enum import table, text
+from docx.shared import Emu, Pt, RGBColor
+from docx.oxml.shared import qn, OxmlElement
+
+
+def create_stylised_document() -> Document:
```
Comment (BognaLew) [inline src/report/report_utils.py:24]: Like `generate()` from report.py.

File: .github/workflows/run_pytest.yml [hunk 8]
```diff
@@ -21,6 +21,6 @@ jobs:
           python -m pip install --upgrade pip  
           if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
       - name: Test with pytest
-        run: coverage run -m pytest test/ -v -s
+        run: coverage run --omit='*/dist-packages/*' -m pytest test/ -v -s
```
Comment (bstrzelecki) [inline .github/workflows/run_pytest.yml]: What is the reason for this change?

File: src/report/report.py (hunk 6 above)
Comment (bstrzelecki) [reply src/report/report.py:17]: not yet resolved

File: src/report/report_utils.py (hunk 7 above)
Comment (bstrzelecki) [reply src/report/report_utils.py:24]: not yet resolved

File: test/report_convert_to_pdf_test.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+def file_cleanup():
+    yield
+    for file in listdir('./'):
+        if file.endswith('.docx'):
+            remove(file)
+        if file.endswith('.pdf'):
+            remove(file)
+
+
+@pytest.mark.skip(reason="Shouldn't be run automatically")
```
Comment (bstrzelecki) [inline test/report_convert_to_pdf_test.py]: Why this shouldn't be run automatically?

File: test/report_convert_to_pdf_test.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+from os.path import exists
+
+import pytest
+from docx import Document
+
+from src.report.report_utils import convert_docx_to_pdf
+
+
+@pytest.fixture(autouse=True)
+def file_cleanup():
```
Comment (bstrzelecki) [inline test/report_convert_to_pdf_test.py]: prefer use of TestCase and thus setUp and tearDown methods

File: src/report/report_utils.py [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+            shading = OxmlElement('w:shd')
+        shading.set(qn('w:fill'), color)
+        properties.append(shading)
+
+
+def convert_docx_to_pdf(docx_path: str, pdf_directory: str) -> str:
+    """
+    Calls LibreOffice to convert provided docx_file to pdf.
+    Converted pdf file is saved in pdf_directory.
+    LibreOffice soffice utility is provided via environmental variable LIBREOFFICE.
```
Comment (bstrzelecki) [inline src/report/report_utils.py:322]: soffice?

Comment (bstrzelecki) [reply src/report/report_utils.py:322]: LIBREOFFICE env var is not documented

File: .github/workflows/run_pytest.yml (hunk 8 above)
Comment (furkacz) [reply .github/workflows/run_pytest.yml]: To prevent pandas and matplotlib dependencies from being included in coverage report, which drops coverage percentage way below minimum required.

File: src/report/report.py (hunk 6 above)
Comment (furkacz) [reply src/report/report.py:17]: No it shouldn't. It's a function that generates this particular document. It serves one purpose only. Splitting it into smaller functions will result in functions that are used once and serve no other purpose.

Comment (furkacz) [reply src/report/report.py:17]: No it shouldn't. It's a function that generates this particular document. It serves one purpose only. Splitting it into smaller functions will result in functions that are used once and serve no other purpose. I've already created functions that can be reused.

File: test/report_convert_to_pdf_test.py (hunk 10 above)
Comment (furkacz) [reply test/report_convert_to_pdf_test.py]: Changed all tests using fixtures to TestCase

File: test/report_convert_to_pdf_test.py (hunk 9 above)
Comment (furkacz) [reply test/report_convert_to_pdf_test.py]: Added libreoffice installation to testing workflow and enabled this test

File: src/report/report_utils.py (hunk 11 above)
Comment (furkacz) [reply src/report/report_utils.py:322]: Updated README to provide more information about setting this environmental variable

File: src/report/report.py (hunk 6 above)
Comment (bstrzelecki) [reply src/report/report.py:17]: This function is far to long, without any sensible labes detailing steps of execution and thus making review process more difficult

**Review by dlydka — APPROVED** [review]

**Review by MikolekN — APPROVED** [review]
